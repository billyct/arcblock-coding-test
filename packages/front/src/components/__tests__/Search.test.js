import {waitFor, fireEvent, render} from '@testing-library/react'

jest.mock('wouter')
const wouter = require('wouter')

afterAll(() => {
  jest.resetModules()
})

describe('test Search Component', () => {

  beforeEach(() => {
    jest.mock('../SearchButton', () => 'search-button-component')
    jest.mock('../SearchInput', () => 'search-input-component')
    jest.mock('../ClearButton', () => 'clear-button-component')
  })

  test('snapshot should correct', () => {
    jest.isolateModules(() => {
      const Search = require('../Search').default
      const {container} = render(<Search className='search_classname'/>)
      expect(container).toMatchSnapshot()
    })
  })

  test('handleChange method should correct', () => {
    jest.unmock('../SearchInput')
    jest.isolateModules(() => {
      const Search = require('../Search').default
      const {getByTestId} = render(<Search/>)

      const input = getByTestId('Search__SearchInput')
      fireEvent.change(input, {
        target: {
          value: '0'
        }
      })

      expect(input).toHaveValue('0')
    })
  })

  test('handleKeyUp method should correct', () => {
    jest.unmock('../SearchInput')
    jest.isolateModules(() => {
      const setLocation = jest.fn()
      wouter.__mockSetLocation(setLocation)

      const Search = require('../Search').default
      const {getByTestId} = render(<Search/>)

      const input = getByTestId('Search__SearchInput')
      fireEvent.keyUp(input, {
        key: 'Enter',
      })

      expect(setLocation).toBeCalledTimes(1)
      expect(setLocation).toBeCalledWith('/block/')
    })
  })

  test(`SearchButton's className should correct`, () => {
    jest.unmock('../SearchInput')
    jest.unmock('../SearchButton')

    jest.isolateModules(async () => {

      const Search = require('../Search').default
      const {getByTestId} = render(<Search/>)

      const input = getByTestId('Search__SearchInput')
      const button = getByTestId('Search__SearchButton')

      fireEvent.change(input, {target: {value: 'not empty'}})
      expect(button).toHaveClass('bg-black text-white')
    })
  })

  test('handleSearch method should correct', () => {
    jest.unmock('../SearchButton')

    jest.isolateModules(() => {
      const setLocation = jest.fn()
      wouter.__mockSetLocation(setLocation)

      const Search = require('../Search').default
      const {getByTestId} = render(<Search/>)

      const button = getByTestId('Search__SearchButton')
      fireEvent.click(button)

      expect(setLocation).toBeCalledTimes(1)
      expect(setLocation).toBeCalledWith('/block/')
    })
  })

  test(`ClearButton's className should correct`, () => {
    jest.unmock('../SearchInput')
    jest.unmock('../ClearButton')

    jest.isolateModules(async () => {

      const Search = require('../Search').default
      const {getByTestId} = render(<Search/>)

      const input = getByTestId('Search__SearchInput')
      const button = getByTestId('Search__ClearButton')

      fireEvent.change(input, {target: {value: 'not empty'}})
      expect(button).toHaveClass('block')

      fireEvent.change(input, {target: {value: ''}})
      expect(button).toHaveClass('hidden')
    })
  })

  test('handleClear method should correct', () => {
    jest.unmock('../SearchInput')
    jest.unmock('../ClearButton')

    jest.isolateModules(async () => {

      const Search = require('../Search').default
      const {getByTestId} = render(<Search/>)

      const input = getByTestId('Search__SearchInput')
      input.value = 'something'

      expect(input).toHaveValue('something')

      fireEvent.click(getByTestId('Search__ClearButton'))

      await waitFor(() => {
        expect(input).toHaveValue('')
        expect(input).toHaveFocus()
      })
    })
  })
})

