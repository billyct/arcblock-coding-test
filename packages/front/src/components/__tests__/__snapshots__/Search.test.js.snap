// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test Search Component snapshot should correct 1`] = `
<div>
  <div
    class="w-full border border-gray-300 leading-loose flex rounded shadow-sm h-12 overflow-hidden hover:shadow search_classname"
  >
    <div
      class="flex-1 relative"
    >
      <search-input-component
        data-testid="Search__SearchInput"
        placeholder="Search a block hash"
        value=""
      />
      <clear-button-component
        classname="hidden"
        data-testid="Search__ClearButton"
      />
    </div>
    <search-button-component
      classname=""
      data-testid="Search__SearchButton"
    />
  </div>
</div>
`;
